// node-misrcon/dist/lib/parsers/parseSysInfoStats.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Name: parseSysInfoStats
 * Description:
 */
var node_misrcon_1 = require("../node-misrcon");
exports.defaultSysInfoState = {
    dn: '',
    pmu: '',
    pmup: '',
    rate: '',
    tpm: '',
    tvm: '',
    up: '',
    upd: '',
    uptime: '',
    vmu: '',
    vmup: ''
};
exports.default = (function (response) {
    if (!response.includes('upd: ')) {
        throw new node_misrcon_1.ParserError('Not a SysInfo Stats Response');
    }
    try {
        return response
            .replace("[CONSOLE] Executing console command 'sysinfo stats'", '')
            .split(',')
            .map(function (item) {
            var _a;
            var splitItems = item.split(':');
            return _a = {},
                _a[splitItems[0].trim()] = splitItems[1].trim(),
                _a;
        })
            .reduce(function (acc, curVal) { return (__assign(__assign({}, acc), curVal)); }, exports.defaultSysInfoState);
    }
    catch (e) {
        throw new node_misrcon_1.ParserError('Not a SysInfo Stats Response');
    }
});
